src/framework/mlt_slices.c

```diff
@@ -306,63 +316,71 @@ void mlt_slices_run( mlt_slices ctx, int jobs, mlt_slices_proc proc, void* cooki
 	pthread_mutex_unlock( &ctx->cond_mutex);
 }
 
-/** Initialize a sliced threading context pool
+/** Get a global shared sliced threading context.
  *
- * \public \memberof mlt_slices_s
- * \param threads number of threads to use for job list
- * \param policy scheduling policy of processing threads
- * \param priority priority value that can be used with the scheduling algorithm
- * \param name name of pool of threads
+ * There are separate contexts for each scheduling policy.
+ *
+ * \private \memberof mlt_slices_s
+ * \param policy the thread scheduling policy needed
  * \return the context pointer
  */
 
-mlt_slices mlt_slices_init_pool( int threads, int policy, int priority, const char* name )
+static mlt_slices mlt_slices_get_global( mlt_schedule_policy policy )
 {
-	mlt_slices ctx = NULL;
-
-	pthread_mutex_lock( &pool_lock );
-
-	/* try to find it by name */
-	if ( name )
+	pthread_mutex_lock( &g_lock );
+	if ( !globals[policy] )
 	{
-		if ( !pool_map )
-		{
-			pool_map = mlt_properties_new();
-			mlt_log_debug( NULL, "%s:%d: pool_map=%p\n", __FUNCTION__, __LINE__, pool_map );
-		}
-		else
-			ctx = (mlt_slices)mlt_properties_get_data( pool_map, name, 0 );
-	}
-
-	if ( !ctx )
-	{
-		ctx = mlt_slices_init( threads, policy, priority );
-		if ( name )
-		{
-			ctx->name = name;
-			mlt_properties_set_data( pool_map, name, ctx, 0, NULL, NULL );
+		int posix_policy;
+		switch (policy) {
+		case mlt_policy_rr:
+			posix_policy = SCHED_RR;
+			break;
+		case mlt_policy_fifo:
+			posix_policy = SCHED_FIFO;
+			break;
+		default:
+			posix_policy = SCHED_OTHER;
 		}
-		mlt_log_debug( NULL, "%s:%d: initialized pool=[%s]\n", __FUNCTION__, __LINE__, name );
-	}
-	else
-	{
-		ctx->ref++;
-		mlt_log_debug( NULL, "%s:%d: reusing pool=[%s]\n", __FUNCTION__, __LINE__, name );
+		globals[policy] = mlt_slices_init( 0, posix_policy, -1 );
+		mlt_factory_register_for_clean_up( globals[policy], (mlt_destructor) mlt_slices_close );
 	}
+	pthread_mutex_unlock( &g_lock );
 
-	pthread_mutex_unlock( &pool_lock );
-
-	return ctx;
+	return globals[policy];
 }
 
 /** Get the number of slices.
  *
  * \public \memberof mlt_slices_s
- * \param ctx context pointer
  * \return the number of slices
  */
 
-int mlt_slices_count(mlt_slices ctx)
+int mlt_slices_count()
+{
+	mlt_slices slices = mlt_slices_get_global( mlt_policy_normal );
+	if (slices)
+		return slices->count;
+	if ((slices = mlt_slices_get_global( mlt_policy_fifo )))
+		return slices->count;
+	if ((slices = mlt_slices_get_global( mlt_policy_rr )))
+		return slices->count;
```

Comment:
Good point; however, your logic requires there to be some previous call such as mlt_slices_run_ to create a context. However, where this is used such as in frei0r, is not the correct place to call run. We could make separate _normal, _rr, and _fifo versions of these since the service will know which policy it wants.